compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -291,8 +291,11 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                         }
                     }
                     None => {
-                        debug!("assemble_unboxed_candidates: closure_kind not yet known");
-                        candidates.vec.push(ClosureCandidate { is_const });
+                        if kind == ty::ClosureKind::FnOnce {
+                            candidates.vec.push(ClosureCandidate { is_const });
+                        } else {
+                            candidates.ambiguous = true;
+                        }
```

Comment:
this slightly weakens inference, does it? we now stall `closure: FnMut` if we don't yet know the closure kind while we could now use that goal to infer e.g. the signature of the closure?

Reply:
Yes, but we always make progress on a corresponding FnOnce projection that gets emitted besides it, which will always make sure we constrain the signature.

I don't immediately see a case where stalling an FnMut or Fn **trait** goal itself will cause inference to be affected negatively, as long as we can make progress on the projection goal and make inference progress on the closure's signature.

Reply:
ah, that's because the only stable usage of the `Fn` traits is via paren sugar which always results in a projection predicate as well?

Reply:
Yes, specifically an `FnOnce` projection goal, which we don't even check the closure kind in (old solver) projection.

Reply:
This *could* result in inference changes if you use the unboxed_closure syntax + raw fn_traits + custom `FnMut` impl with where clauses that affect inference, but that's 1. unstable, and 2. not something we can support in the new trait solver, so I don't feel particularly symathetic to that use-case even if it does exist (which I'm almost certain it does not).